README.md

# Sentiment Analysis
![Test Suite](https://github.com/paluchasz/sentiment_analysis/actions/workflows/check_pr.yaml/badge.svg?branch=main)

Using transformers to predict the sentiment of movie reviews from the Stanford dataset available [here](https://ai.stanford.edu/~amaas/data/sentiment/).


## Setup

We use poetry to manage the virtual env and dependency management.

- Install poetry from [here](https://python-poetry.org/docs/)
- From project root run
```
poetry install
```
this will create a virtual env in a `.venv` folder and install all dependencies into it.
- To activate a shell in the venv run
```
poetry shell
```

## Training scripts
To run the scripts the sample_data and sample env file can be used. For the full dataset download from
[here](https://ai.stanford.edu/~amaas/data/sentiment/).

The 3 scripts are defined as executables in the `pyproject.toml` file and if this repository is installed
as a package with either `poetry install` or `pip install .` the scripts can be called with simply: `get_train_test_data`,
`train` and `predict_and_evaluate`.


CHANGELOG.md
# Changelog

- This file provides a full account of all changes to the model/api
- This file is manually updated as we merge PRs. Simply add a heading at the top of this lsit with a link to your PR in it with a _short_ list of changes

## Cleanup code (https://github.com/paluchasz/sentiment_analysis/pull/5)
- Make sure code is working, clean up code into initial separate sub-modules: getting train/test data, training and evaluation.

## Extras (https://github.com/paluchasz/sentiment_analysis/pull/4)

- Some extra commit hooks
- Our first badges! One for doc string coverage and one for status of tests workflow on main branch
## Poetry (https://github.com/paluchasz/sentiment_analysis/pull/3)

- Set up venv and dependency management using poetry
- Add poetry pre commit hooks

## Commit hooks (https://github.com/paluchasz/sentiment_analysis/pull/2)
- Add pre commit config for code styling on commits

## Setup (https://github.com/paluchasz/sentiment_analysis/pull/1)
- Add this changlog
- Add CODEOWNERS FILE
- Add PR template
- Add .gitignore
- A hello world github actions pr check workflow


sentiment_analysis/test.py
# test
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split

from sentiment_analysis import DATA_DIR, get_data

v = tf.Variable(1)


@tf.function
def f(x):
    """A doc string to get interrogate commit hook passing"""
    ta = tf.TensorArray(tf.float32, size=0, dynamic_size=True)
    for i in tf.range(x):
        v.assign_add(i)
        ta = ta.write(i, 0.123)
    return ta.stack() / 3


a = f(5)

training_data = get_data(from_file=DATA_DIR / "training_data.json")
print(len(training_data))

np.random.seed(1)
np.random.shuffle(training_data)
X_train, y_train = [list(x) for x in zip(*[[t["text"], t["label"]] for t in training_data])]

train_texts, val_texts, train_labels, val_labels = train_test_split(X_train, y_train, test_size=0.2)
a = 1


sentiment_analysis/scripts/train.py
import json
import time
from pathlib import Path
from typing import Optional

import numpy as np
import tensorflow as tf
from dotenv import load_dotenv
from pydantic import BaseSettings
from pydantic.class_validators import root_validator
from sklearn.model_selection import train_test_split
from transformers import DistilBertTokenizer, TFDistilBertForSequenceClassification

GPU_AVAILABLE = tf.test.is_gpu_available()

load_dotenv(".env")  # need to run script from root of project to read this


class EnvVars(BaseSettings):
    """Pydantic data class to hold environment variables"""

    TRAIN_DATA_PATH: Path
    EPOCHS: int = 1
    BATCH_SIZE: int = 64 if GPU_AVAILABLE else 1
    VAL_BATCH_SIZE: int = 16 if GPU_AVAILABLE else 1
    SAVE_MODEL_DIR: Optional[Path] = None

    @root_validator
    def _convert_relative_to_absolute_paths(cls, values):
        values["TRAIN_DATA_PATH"] = values["TRAIN_DATA_PATH"].resolve()
        if values["SAVE_MODEL_DIR"] is not None:
            values["SAVE_MODEL_DIR"] = values["SAVE_MODEL_DIR"].resolve()
        return values


ENV_VARS = EnvVars()


def custom_training(
    model, tokenizer, x, y, epochs=1, batch_size=64, val_batch_size=16, save_model_dir=None, stats_dir=None, val_step_freq=None
):
    """Implements a custom training loop so can validate more often"""
    start_time = time.time()
    train_texts, val_texts, train_labels, val_labels = train_test_split(x, y, test_size=0.05)

    train_encodings = tokenizer(train_texts, truncation=True, padding=True)
    val_encodings = tokenizer(val_texts, truncation=True, padding=True)
    print("Found all tokens in {} min".format((time.time() - start_time) / 60))

    # Make tf dataset objects
    train_dataset = tf.data.Dataset.from_tensor_slices((dict(train_encodings), train_labels))
    train_dataset = train_dataset.shuffle(1000).batch(batch_size)
    val_dataset = tf.data.Dataset.from_tensor_slices((dict(val_encodings), val_labels))
    val_dataset = val_dataset.shuffle(1000).batch(val_batch_size)

    optimizer = tf.keras.optimizers.Adam(learning_rate=5e-5)
    loss_func = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

    train_losses = []
    val_losses = []
    steps = int(np.ceil(len(train_dataset)))
    if not val_step_freq:
        val_step_freq = max(1, int(steps / 10))
    for epoch in range(epochs):
        print("Starting epoch {}/{}".format(epoch + 1, epochs))
        for step, (x_batch, y_batch) in enumerate(train_dataset):
            step_time = time.time()
            train_loss = train_step(model, x_batch, y_batch, loss_func, optimizer)
            print("Training loss: {}".format(train_loss))
            train_losses.append([epoch, step, steps, float(train_loss.numpy())])

            if step % val_step_freq == 0:
                val_time = time.time()

                # Validate every certain number of steps
                total_val_loss = validation_step(model, val_dataset, loss_func) / len(val_dataset)
                val_losses.append([epoch, step, steps, float(total_val_loss.numpy())])
                print("Validation loss: {}, time {}".format(total_val_loss, time.time() - val_time))
            print("Step {}/{} Step time: {}".format(step + 1, steps, time.time() - step_time))

    print("Training time: {}".format(time.time() - start_time))
    if save_model_dir:
        print(f"Saving model in {save_model_dir}")
        model.save_pretrained(save_model_dir)
        tokenizer.save_pretrained(save_model_dir)
    if stats_dir:
        print("Saving stats")
        json.dump(
            {"train_losses": train_losses, "val_losses": val_losses},
            open(stats_dir, "w"),
        )


@tf.function
def validation_step(model, val_dataset, loss_func):
    """Performs one pass through validation set"""
    total_loss = tf.constant(0, dtype=tf.float32)
    for i, (x_batch_val, y_batch_val) in enumerate(val_dataset):
        val_logits = model(x_batch_val, training=False)
        val_loss = loss_func(y_batch_val, val_logits.logits)
        total_loss += val_loss

    return total_loss  # tf.math.divide(total_loss, tf.constant(len(val_dataset), dtype=tf.float32))


@tf.function  # comment out to run in debug
def train_step(model, x, y, loss_func, optimizer):
    """Implements one training step. Use decorator for fast performance using graphs. Seems to halve the train time!"""
    # Open a GradientTape to records operations during the forward pass which enables auto-differentiation
    with tf.GradientTape() as tape:
        # Run forward pass, calculate loss
        train_logits = model(x, training=True)
        train_loss = loss_func(y, train_logits.logits)
    # Use the tape tp automatically retrieve the gradients of the trainable variables with respect to the loss
    grads = tape.gradient(train_loss, model.trainable_weights)
    # Run one step of gradient descent by updating the values of the variables to minimise the loss
    optimizer.apply_gradients(zip(grads, model.trainable_weights))

    return train_loss


def main() -> None:
    """Train model"""
    with open(ENV_VARS.TRAIN_DATA_PATH, "r") as file:
        train_data = json.load(file)

    model = TFDistilBertForSequenceClassification.from_pretrained("distilbert-base-uncased")
    tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")

    x_train, y_train = [list(x) for x in zip(*[[t["text"], t["label"]] for t in train_data])]

    custom_training(
        model,
        tokenizer,
        x_train,
        y_train,
        epochs=ENV_VARS.EPOCHS,
        batch_size=ENV_VARS.BATCH_SIZE,
        val_batch_size=ENV_VARS.VAL_BATCH_SIZE,
        save_model_dir=ENV_VARS.SAVE_MODEL_DIR,
    )


if __name__ == "__main__":
    main()


sentiment_analysis/model_analysis.py
import json

import plotly.graph_objects as go

from sentiment_analysis import DATA_DIR


def model_analysis():
    """doc string to get interrogate to pass"""
    with open(DATA_DIR / "stats/my_sentiment_model_valid_e1b64_losses.json", "r") as file:
        losses = json.load(file)

    train_losses = [[e[0] * e[2] + e[1], e[3]] for e in losses["train_losses"] if e[1] % 5 == 0]
    val_losses = [[e[0] * e[2] + e[1], e[3]] for e in losses["val_losses"]]
    train_x, train_y = zip(*train_losses)
    val_x, val_y = zip(*val_losses)

    layout = go.Layout(title="Losses per step", xaxis={"title": "Step"}, yaxis={"title": "Loss"})
    fig = go.Figure(layout=layout)
    fig.add_trace(go.Scatter(x=train_x, y=train_y, name="train error", mode="lines+markers"))
    fig.add_trace(go.Scatter(x=val_x, y=val_y, name="val error", mode="lines+markers"))
    fig.show()
    a = 1
    print(a)


model_analysis()


sentiment_analysis/tf_gradient_tape_example.py
import sys
import time

import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.layers import (
    Activation,
    BatchNormalization,
    Conv2D,
    Dense,
    Dropout,
    Flatten,
    MaxPooling2D,
)
from tensorflow.keras.losses import categorical_crossentropy
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical


def build_model(width, height, depth, classes):
    """A doc string to get interrogate commit hook passing"""

    # initialize the input shape and channels dimension to be
    # "channels last" ordering
    inputShape = (height, width, depth)
    chanDim = -1
    # build the model using Keras' Sequential API
    model = Sequential(
        [
            # CONV => RELU => BN => POOL layer set
            Conv2D(16, (3, 3), padding="same", input_shape=inputShape),
            Activation("relu"),
            BatchNormalization(axis=chanDim),  # normalises the channel axis
            MaxPooling2D(
                pool_size=(2, 2)
            ),  # we down sample the feature map which recorded the precise position of features in input
            # Helps with changes like cropping, rotating ,shifting of image which would result in different feature maps.
            # (CONV => RELU => BN) * 2 => POOL layer set
            Conv2D(32, (3, 3), padding="same"),
            Activation("relu"),
            BatchNormalization(axis=chanDim),
            Conv2D(32, (3, 3), padding="same"),
            Activation("relu"),
            BatchNormalization(axis=chanDim),
            MaxPooling2D(pool_size=(2, 2)),
            # (CONV => RELU => BN) * 3 => POOL layer set
            Conv2D(64, (3, 3), padding="same"),
            Activation("relu"),
            BatchNormalization(axis=chanDim),
            Conv2D(64, (3, 3), padding="same"),
            Activation("relu"),
            BatchNormalization(axis=chanDim),
            Conv2D(64, (3, 3), padding="same"),
            Activation("relu"),
            BatchNormalization(axis=chanDim),
            MaxPooling2D(pool_size=(2, 2)),
            # first (and only) set of FC => RELU layers
            Flatten(),
            Dense(256),
            Activation("relu"),
            BatchNormalization(),
            Dropout(0.5),
            # softmax classifier
            Dense(classes),
            Activation("softmax"),
        ]
    )
    # return the built model to the calling function
    return model


def step(X, y):
    """A doc string to get interrogate commit hook passing"""
    # keep track of our gradients
    with tf.GradientTape() as tape:
        # make a prediction using the model and then calculate the
        # loss
        pred = model(X)
        loss = categorical_crossentropy(y, pred)
    # calculate the gradients using our tape and then update the
    # model weights
    grads = tape.gradient(loss, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))


# initialize the number of epochs to train for, batch size, and
# initial learning rate
EPOCHS = 25
BS = 64
INIT_LR = 1e-3

# load the MNIST dataset
print("[INFO] loading MNIST dataset...")
((trainX, trainY), (testX, testY)) = mnist.load_data()

# add a channel dimension to every image in the dataset, then scale
# the pixel intensities to the range [0, 1]
# Adding channel simply adds an extra dimension with information about each colour of a pixel. E.g array of shape
# (100, 150) for width and height of pixel is converted to either (100, 150, 1) greyscale or (100, 150, 3) colour scale.
# (3 components for red, blue and green)
trainX = np.expand_dims(trainX, axis=-1)
testX = np.expand_dims(testX, axis=-1)
trainX = trainX.astype("float32") / 255.0
testX = testX.astype("float32") / 255.0

# one-hot encode the labels
trainY = to_categorical(trainY, num_classes=10)  # converts each label like 5 to [0,0,0,0,0,1,0,0,0,0]
testY = to_categorical(testY, num_classes=10)

# build our model and initialize our optimizer
print("[INFO] creating model...")
model = build_model(28, 28, 1, 10)
opt = Adam(lr=INIT_LR, decay=INIT_LR / EPOCHS)


# compute the number of batch updates per epoch
numUpdates = int(trainX.shape[0] / BS)
# loop over the number of epochs
for epoch in range(0, EPOCHS):
    # show the current epoch number
    print("[INFO] starting epoch {}/{}...".format(epoch + 1, EPOCHS), end="")
    sys.stdout.flush()
    epochStart = time.time()
    # loop over the data in batch size increments
    for i in range(0, numUpdates):
        # determine starting and ending slice indexes for the current
        # batch
        start = i * BS
        end = start + BS
        # take a step
        step(trainX[start:end], trainY[start:end])
    # show timing information for the epoch
    epochEnd = time.time()
    elapsed = (epochEnd - epochStart) / 60.0
    print("took {:.4} minutes".format(elapsed))


# in order to calculate accuracy using Keras' functions we first need
# to compile the model
model.compile(optimizer=opt, loss=categorical_crossentropy, metrics=["acc"])
# now that the model is compiled we can compute the accuracy
(loss, acc) = model.evaluate(testX, testY)
print("[INFO] test accuracy: {:.4f}".format(acc))


sentiment_analysis/scripts/get_train_test_data.py
from __future__ import annotations

"""
To use this script download raw data from https://ai.stanford.edu/~amaas/data/sentiment/ and specify path to the
aclImdb folder. If installed with poetry can simply run `get_train_test_data`.
"""
import json
from pathlib import Path

import numpy as np
from dotenv import load_dotenv
from pydantic import BaseSettings, root_validator

load_dotenv(".env")


class EnvVars(BaseSettings):
    """Pydantic data class to hold environment variables"""

    RAW_DATA_DIR: Path
    TRAIN_TEST_DIR: Path

    @root_validator
    def _convert_relative_to_absolute_paths(cls, values):
        values["RAW_DATA_DIR"] = values["RAW_DATA_DIR"].resolve()
        values["TRAIN_TEST_DIR"] = values["TRAIN_TEST_DIR"].resolve()
        return values


ENV_VARS = EnvVars()


def convert_raw_data(dir_path: Path) -> list[dict[str, str | int]]:
    """Convert original raw data format into train and test data files"""
    print("Converting raw data")
    examples = []
    for label_dir in ["pos", "neg"]:
        for example in Path(dir_path / label_dir).glob("*txt"):
            ex_id, ex_rating = example.stem.split("_")
            with open(example, "r") as file:
                text = file.read()
            examples.append(
                {
                    "text": text,
                    "id": int(ex_id),
                    "ex_rating": int(ex_rating),
                    "label": 1 if label_dir == "pos" else 0,
                }
            )
    np.random.shuffle(examples)
    return examples


def main() -> None:
    """Load road data and save as single train and test files"""
    train_data = convert_raw_data(Path(ENV_VARS.RAW_DATA_DIR / "aclImdb/train"))
    test_data = convert_raw_data(Path(ENV_VARS.RAW_DATA_DIR / "aclImdb/test"))

    print(f"Saving train and test data in {ENV_VARS.TRAIN_TEST_DIR}")
    with open(ENV_VARS.TRAIN_TEST_DIR / "train.json", "w") as file:
        json.dump(train_data, file)
    with open(ENV_VARS.TRAIN_TEST_DIR / "test.json", "w") as file:
        json.dump(test_data, file)


if __name__ == "__main__":
    """Doc string for check to pass. Not needed?"""
    main()


sentiment_analysis/scripts/predict_and_evaluate.py
import json
import time
from pathlib import Path

import numpy as np
import tensorflow as tf
from dotenv import load_dotenv
from pydantic.class_validators import root_validator
from pydantic.env_settings import BaseSettings
from transformers import AutoTokenizer, TFAutoModelForSequenceClassification

GPU_AVAILABLE = tf.test.is_gpu_available()

load_dotenv(".env")  # need to run script from root of project to read this


class EnvVars(BaseSettings):
    """Pydantic data class to hold environment variables"""

    TEST_DATA_PATH: Path
    TRAINED_MODEL_DIR: Path
    BATCH_SIZE: int = 64 if GPU_AVAILABLE else 1

    @root_validator
    def _convert_relative_to_absolute_paths(cls, values):
        values["TEST_DATA_PATH"] = values["TEST_DATA_PATH"].resolve()
        values["TRAINED_MODEL_DIR"] = values["TRAINED_MODEL_DIR"].resolve()
        return values


ENV_VARS = EnvVars()


def predict_and_evaluate(model, tokenizer, x, y, batch_size=64):
    """Predict and evaluate on the test set"""
    predictions = []
    batches_num = int(np.math.ceil(len(x) / batch_size))
    print("Number of batches {} of size {}".format(batches_num, batch_size))
    start_time = time.time()
    for i in range(batches_num):
        print("Batch {},".format(i), end=" ")
        batch_start_time = time.time()
        input_tokens = tokenizer(
            x[i * batch_size : min((i + 1) * batch_size, len(x))],
            padding=True,
            truncation=True,
            max_length=512,
            return_tensors="tf",
        )
        print("Tokenizer time: {},".format(time.time() - batch_start_time), end=" ")
        tf_outputs = model(input_tokens)
        predictions += tf.nn.softmax(tf_outputs.logits, axis=-1).numpy().tolist()
        print("Batch time: {}".format(time.time() - batch_start_time))

    print("Time taken: {} minutes ".format((time.time() - start_time) / 60))
    print("Predictions: ", predictions)
    predictions = [1 if p[1] > 0.5 else 0 for p in predictions]
    print("Predictions: ", predictions)
    print("Actual:      ", y)
    accuracy = sum([1 for i in range(len(predictions)) if predictions[i] == y[i]]) / len(predictions)
    print("Accuracy: {}".format(accuracy))
    return predictions


def main() -> None:
    """Evaluate the current model on the test set"""
    with open(ENV_VARS.TEST_DATA_PATH, "r") as file:
        test_data = json.load(file)
    x_test, y_test = [list(x) for x in zip(*[[t["text"], t["label"]] for t in test_data])]

    tokenizer = AutoTokenizer.from_pretrained(ENV_VARS.TRAINED_MODEL_DIR)
    model = TFAutoModelForSequenceClassification.from_pretrained(ENV_VARS.TRAINED_MODEL_DIR)

    predict_and_evaluate(model, tokenizer, x_test, y_test)


if __name__ == "__main__":
    main()


sentiment_analysis/sentiment_analysis.py
import json
import time
from pathlib import Path

import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds
from sklearn.model_selection import train_test_split
from transformers import (
    AutoTokenizer,
    DistilBertTokenizer,
    TFAutoModelForSequenceClassification,
    TFDistilBertForSequenceClassification,
    glue_convert_examples_to_features,
    pipeline,
)

DATA_DIR = Path("/Users/szymon.palucha/Desktop/sentiment_analysis/data/aclImdb")


def get_data(test=False, from_file=None):
    """Gets train data by default unless test=True. Loads from one file if path passed in."""
    if from_file:
        with open(from_file, "r") as file:
            data = json.load(file)
        return data

    dir_path = DATA_DIR / "test" if test else DATA_DIR / "train"

    examples = []
    for label_dir in ["pos", "neg"]:
        for example in (dir_path / label_dir).glob("*txt"):
            ex_id, ex_rating = example.stem.split("_")
            with open(example, "r") as file:
                text = file.read()
            examples.append(
                {
                    "text": text,
                    "id": int(ex_id),
                    "ex_rating": int(ex_rating),
                    "label": 1 if label_dir == "pos" else 0,
                }
            )

    return examples


def predict(x, y, batch_size=100):
    """Using standard pipeline, disadvantage: not sure how many characters to limit to so < 512 tokens."""
    classifier = pipeline("sentiment-analysis")
    predictions = []
    batches_num = int(np.math.ceil(len(x) / batch_size))
    print("Number of batches {} of size {}".format(batches_num, batch_size))
    start_time = time.time()
    for i in range(batches_num):
        print("Batch {}".format(i))
        # Limit to 1000 characters
        predictions += classifier([d[: min(1000, len(d))] for d in x[i * batch_size : min((i + 1) * batch_size, len(x))]])

    print("Time taken {} minutes ".format((time.time() - start_time) / 60))
    predictions = [1 if p["label"] == "POSITIVE" else 0 for p in predictions]
    accuracy = sum([1 for i in range(len(predictions)) if predictions[i] == y[i]]) / len(predictions)
    print("Accuracy: {}".format(accuracy))
    return predictions


def predict_with_tokenizer(x, y, batch_size=100):
    """Need to predict in batches or will run out of memory"""
    predictions = []
    batches_num = int(np.math.ceil(len(x) / batch_size))
    print("Number of batches {} of size {}".format(batches_num, batch_size))
    start_time = time.time()
    for i in range(batches_num):
        print("Batch {},".format(i), end=" ")
        batch_start_time = time.time()
        input_tokens = tokenizer(
            x[i * batch_size : min((i + 1) * batch_size, len(x))],
            padding=True,
            truncation=True,
            max_length=512,
            return_tensors="tf",
        )
        print("Tokenizer time: {},".format(time.time() - batch_start_time), end=" ")
        tf_outputs = tf_model(input_tokens)
        predictions += tf.nn.softmax(tf_outputs.logits, axis=-1).numpy().tolist()
        print("Batch time: {}".format(time.time() - batch_start_time))

    print("Time taken: {} minutes ".format((time.time() - start_time) / 60))
    print("Predictions: ", predictions)
    predictions = [1 if p[1] > 0.5 else 0 for p in predictions]
    print("Predictions: ", predictions)
    print("Actual:      ", y)
    accuracy = sum([1 for i in range(len(predictions)) if predictions[i] == y[i]]) / len(predictions)
    print("Accuracy: {}".format(accuracy))
    return predictions


def tokenizer_example():
    """Simple example, define tokenizer and model in main"""
    inputs = tokenizer(
        "This is a small positive example",
        padding=True,
        truncation=True,
        max_length=512,
        return_tensors="tf",
    )
    # Need to pad and truncate to 512 tokens, return_tensors to not give lists for input_ids/attention and mask and so
    # it can be passed to the model. Attention mask is 0 for pad tokens - to skip them from self attention
    # To pass in a pair of sentences (like question answering) pass in [[Sentence A, Sentence B], ...], or if just a
    # single pair pass in as two arguments, (a list is interpreted as a batch), so can also pass in two lists as two
    # arguments: one for all first sentences and one for all second sentences. Token_type_ids will indicate which
    # sentence each input sequence belongs too
    tf_outputs = tf_model(inputs)
    # First number in tuple is the score of negative, second is the score of positive sentiment, pass through softmax
    tf_predictions = tf.nn.softmax(tf_outputs[0], axis=-1)
    # Or give label and will softmax loss already, can also output states
    tf_outputs_2 = tf_model(
        inputs,
        labels=tf.constant([0]),
        output_hidden_states=True,
        output_attentions=True,
    )
    return tf_outputs_2, tf_predictions


def mrpc_example():
    """This is an example of Microsoft research paraphrase corpus, predicts whether two senteces are paraphrases"""
    original = "Giraffes like Acacia leaves and hay, and they can consume 75 pounds of food a day."
    paraphrase = "A giraffe can eat up to 75 pounds of Acacia leaves and hay daily."
    tokenizer = AutoTokenizer.from_pretrained("bert-base-cased-finetuned-mrpc")
    model = TFAutoModelForSequenceClassification.from_pretrained("bert-base-cased-finetuned-mrpc")
    tokens = tokenizer(original, paraphrase, return_tensors="tf")
    result = model(tokens)
    probabilities = tf.nn.softmax(result.logits)  # 1st: prob of not paraphrase, 2nd: prob of paraphrase
    return probabilities


def test_fine_tuned_model(model_dir):
    """A doc string to get interrogate commit hook passing"""

    model = TFDistilBertForSequenceClassification.from_pretrained(model_dir)
    tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")

    tokens = tokenizer(
        "This is a positive sentence",
        truncation=True,
        padding=True,
        return_tensors="tf",
    )
    result = model(tokens)
    probabilities = tf.nn.softmax(result.logits)
    return probabilities


def tf_dataset_example():
    """A doc string to get interrogate commit hook passing"""

    # Dataset example -------
    tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
    data = tfds.load("glue/mrpc")
    # To view dataset object iterate through e.g. [x for x in data['train']]
    train_dataset_0 = glue_convert_examples_to_features(data["train"], tokenizer, max_length=128, task="mrpc")
    train_dataset_0 = train_dataset_0.shuffle(100).batch(32).repeat(2)
    # can view with [x for x in train_dataset]


def standard_training(x, y, epochs=1, batch_size=64, save_model_dir=None, stats_dir=None):
    """A doc string to get interrogate commit hook passing"""

    start_time = time.time()
    # Make a validation set
    train_texts, val_texts, train_labels, val_labels = train_test_split(x, y, test_size=0.2)

    train_encodings = tokenizer(train_texts, truncation=True, padding=True)
    val_encodings = tokenizer(val_texts, truncation=True, padding=True)
    print("Found all tokens in {} min".format((time.time() - start_time) / 60))

    # Make tf dataset objects
    train_dataset = tf.data.Dataset.from_tensor_slices((dict(train_encodings), train_labels))
    train_dataset = train_dataset.shuffle(1000).batch(batch_size)
    val_dataset = tf.data.Dataset.from_tensor_slices((dict(val_encodings), val_labels))
    val_dataset = val_dataset.shuffle(1000).batch(batch_size)

    optimizer = tf.keras.optimizers.Adam(learning_rate=5e-5)
    model.compile(optimizer=optimizer, loss=model.compute_loss)
    print("Compiled model {}".format((time.time() - start_time) / 60))

    # TensorBoard callback can save train losses after every batch. In terminal $ tensorboard --logdir=path_to_your_logs
    # tf_callback = tf.keras.callbacks.TensorBoard(log_dir=DATA_DIR / 'logs', update_freq='batch', write_images=True,
    #                                             histogram_freq=1, embeddings_freq=1)
    history = model.fit(
        train_dataset, validation_data=val_dataset, epochs=epochs, batch_size=batch_size
    )  # callbacks=[tf_callback])
    print("Training took {}".format(time.time() - start_time))
    if save_model_dir:
        print("Saving model")
        model.save_pretrained(save_model_dir)
    if stats_dir:
        json.dump(history.history, open(stats_dir, "w"))


def custom_training(
    x,
    y,
    epochs=1,
    batch_size=64,
    val_batch_size=16,
    save_model_dir=None,
    stats_dir=None,
    val_step_freq=None,
):
    """Implements a custom training loop so can validate more often"""
    start_time = time.time()
    train_texts, val_texts, train_labels, val_labels = train_test_split(x, y, test_size=0.05)

    train_encodings = tokenizer(train_texts, truncation=True, padding=True)
    val_encodings = tokenizer(val_texts, truncation=True, padding=True)
    print("Found all tokens in {} min".format((time.time() - start_time) / 60))

    # Make tf dataset objects
    train_dataset = tf.data.Dataset.from_tensor_slices((dict(train_encodings), train_labels))
    train_dataset = train_dataset.shuffle(1000).batch(batch_size)
    val_dataset = tf.data.Dataset.from_tensor_slices((dict(val_encodings), val_labels))
    val_dataset = val_dataset.shuffle(1000).batch(val_batch_size)

    optimizer = tf.keras.optimizers.Adam(learning_rate=5e-5)
    loss_func = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

    train_losses = []
    val_losses = []
    steps = int(np.ceil(len(train_dataset)))
    if not val_step_freq:
        val_step_freq = int(steps / 10)
    for epoch in range(epochs):
        print("Starting epoch {}/{}".format(epoch + 1, epochs))
        for step, (x_batch, y_batch) in enumerate(train_dataset):
            step_time = time.time()
            train_loss = train_step(x_batch, y_batch, loss_func, optimizer)
            print("Training loss: {}".format(train_loss))
            train_losses.append([epoch, step, steps, float(train_loss.numpy())])

            if step % val_step_freq == 0:
                val_time = time.time()

                # Validate every certain number of steps
                total_val_loss = validation_step(val_dataset, loss_func) / len(val_dataset)
                val_losses.append([epoch, step, steps, float(total_val_loss.numpy())])
                print("Validation loss: {}, time {}".format(total_val_loss, time.time() - val_time))
            print("Step {}/{} Step time: {}".format(step + 1, steps, time.time() - step_time))

    print("Training time: {}".format(time.time() - start_time))
    if save_model_dir:
        print("Saving model")
        model.save_pretrained(save_model_dir)
    if stats_dir:
        print("Saving stats")
        json.dump(
            {"train_losses": train_losses, "val_losses": val_losses},
            open(stats_dir, "w"),
        )


@tf.function
def validation_step(val_dataset, loss_func):
    """A doc string to get interrogate commit hook passing"""

    total_loss = tf.constant(0, dtype=tf.float32)
    for i, (x_batch_val, y_batch_val) in enumerate(val_dataset):
        val_logits = model(x_batch_val, training=False)
        val_loss = loss_func(y_batch_val, val_logits.logits)
        total_loss += val_loss

    return total_loss  # tf.math.divide(total_loss, tf.constant(len(val_dataset), dtype=tf.float32))


@tf.function  # comment out to run in debug
def train_step(x, y, loss_func, optimizer):
    """Implements one training step. Use decorator for fast performance using graphs. Seems to halve the train time!"""
    # Open a GradientTape to records operations during the forward pass which enables auto-differentiation
    with tf.GradientTape() as tape:
        # Run forward pass, calculate loss
        train_logits = model(x, training=True)
        train_loss = loss_func(y, train_logits.logits)
    # Use the tape tp automatically retrieve the gradients of the trainable variables with respect to the loss
    grads = tape.gradient(train_loss, model.trainable_weights)
    # Run one step of gradient descent by updating the values of the variables to minimise the loss
    optimizer.apply_gradients(zip(grads, model.trainable_weights))

    return train_loss


if __name__ == "__main__":
    training_data = get_data(from_file=DATA_DIR / "training_data.json")
    test_data = get_data(from_file=DATA_DIR / "test_data.json")

    np.random.seed(1)
    np.random.shuffle(training_data)
    x_train_, y_train_ = [list(x) for x in zip(*[[t["text"], t["label"]] for t in training_data])]

    x_test_, y_test_ = [list(x) for x in zip(*[[t["text"], t["label"]] for t in test_data])]

    # Using existing pipeline classifier ---------------
    # predict(x_test, y_test)

    # Using tokenizer - equivalent to above  -----------
    model_name = "distilbert-base-uncased-finetuned-sst-2-english"
    # tf_model = TFAutoModelForSequenceClassification.from_pretrained(model_name)
    # tokenizer = AutoTokenizer.from_pretrained(model_name)
    # # For this model the Auto Model/Tokenizer is equivalent to:
    # tf_model = TFDistilBertForSequenceClassification.from_pretrained(model_name)
    # # A linear layer on top of the pooled output (of DistilBert)
    tokenizer = DistilBertTokenizer.from_pretrained(model_name)

    tokenizer_example()
    # predict_with_tokenizer(x_test, y_test)

    test = False
    if test:
        # test_fine_tuned_model(DATA_DIR / 'my_sentiment_model/')
        tf_model = TFDistilBertForSequenceClassification.from_pretrained(DATA_DIR / "my_sentiment_model_e1b64")
        tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
        predict_with_tokenizer(x_test_[:10], y_test_[:10])

    fine_tune_model = True
    if fine_tune_model:
        # Fine tuning ------------------------------------
        model = TFDistilBertForSequenceClassification.from_pretrained("distilbert-base-uncased")
        tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")

        # custom_training(x_train_, y_train_, save_model_dir=DATA_DIR / 'my_sentiment_model_valid_2',
        #                 stats_dir=DATA_DIR / 'my_sentiment_model_valid_2_losses.json', batch_size=16)
        custom_training(x_train_[:10], y_train_[:10], batch_size=2, val_step_freq=2)

        # standard_training(x_train_[:50], y_train_[:50])

    a = 1


# Standard sentiment-analysis pipeline: accuracy = 0.867 (1000 char limit)
# Using tokenizer : Time taken: 142.12985572020213 minutes, Accuracy: 0.88871485943775

# Fined_tuned, epochs: 3, batch_size=16, steps_per_epoch=50; Time: 153min, Accuracy: 0.861285
# Fine_tuned, epochs: 1, batch_size=64, steps_per_epoch=None (ie 25k/64=391), Time >2h Accuracy: 0.91376

#   1/157 [..............................] - ETA: 1:08:25 - loss: 0.6964
#   2/1563 [..............................] - ETA: 6:41:59 - loss: 0.6910
